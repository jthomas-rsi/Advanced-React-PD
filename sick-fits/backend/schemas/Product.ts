/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/unbound-method */
import { text, select, integer, relationship } from '@keystone-next/fields';
import { list } from '@keystone-next/keystone/schema';
import { rules, isSignedIn } from '../access';

export const Product = list({
  access: {
    create: isSignedIn,
    read: rules.canReadProducts,
    update: rules.canManageProducts,
    delete: rules.canManageProducts,
  },
  fields: {
    // list the fields to be rendered in the UI of the Keystone admin for created products
    name: text({ isRequired: true }), // text input field
    description: text({
      // text area field
      ui: {
        displayMode: 'textarea',
      },
    }),
    photo: relationship({
      // relationship field
      ref: 'ProductImage.product', // relationship to ProductImage schema
      ui: {
        displayMode: 'cards',
        cardFields: ['image', 'altText'],
        inlineCreate: { fields: ['image', 'altText'] },
        inlineEdit: { fields: ['image', 'altText'] },
      },
    }),
    status: select({
      // dropdown field
      options: [
        { label: 'Draft', value: 'DRAFT' },
        { label: 'Available', value: 'AVAILABLE' },
        { label: 'Unavailable', value: 'UNAVAILABLE' },
      ],
      defaultValue: 'DRAFT', // setting default value of sataus dropdown to DRAFT
      ui: {
        displayMode: 'segmented-control',
        createView: { fieldMode: 'hidden' },
      },
    }),
    price: integer(), // integer input field
    user: relationship({
      // when the product is created it will be associated with the user that created it
      ref: 'User.products',
      defaultValue: ({ context }) => ({
        connect: { id: context.session.itemId },
      }),
    }),
  },
});
