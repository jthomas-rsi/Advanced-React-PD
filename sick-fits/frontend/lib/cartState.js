// This is our cart context state. We will keep all of our cart state and functionality in this file.
// We will also create a custom hook to access the cart state and functionality from any component in our app.

import { createContext, useContext, useState } from 'react';
import PropTypes from 'prop-types';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

// This is our own custom provider. Storing cart data (state) and functionality (updaters)
const CartStateProvider = ({ children }) => {
  const [isCartOpen, setIsCartOpen] = useState(false);

  // helper functions to update cart state
  const toggleCart = () => setIsCartOpen(!isCartOpen);

  const closeCart = () => setIsCartOpen(false);

  const openCart = () => setIsCartOpen(true);

  return (
    <LocalStateProvider
      value={{
        isCartOpen,
        setIsCartOpen,
        toggleCart,
        closeCart,
        openCart,
      }}
    >
      {children}
    </LocalStateProvider>
  );
};

CartStateProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

// custom hook for accessing the cart local state
const useCart = () => {
  const cart = useContext(LocalStateContext);
  return cart;
};

export { CartStateProvider, useCart };
