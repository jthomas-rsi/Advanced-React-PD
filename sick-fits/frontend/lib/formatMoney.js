const formatMoney = (amount = 0) => {
  // options object to pass to the formatter method
  const options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2,
  };

  // check if it's a clean dollar amount using modulus
  // if it is, we don't want to show decimal places
  if (amount % 100 === 0) {
    options.minimumFractionDigits = 0;
  }

  const formatter = new Intl.NumberFormat('en-US', options);

  // return the formatted dollar amount divided by 100
  return formatter.format(amount / 100);
};

export default formatMoney;
