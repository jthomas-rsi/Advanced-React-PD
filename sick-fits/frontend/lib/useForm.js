import { useState } from 'react';

// this function is a hook to handle from data without using third party library
const useForm = (initial = {}) => {
  // create a state object for our inputs
  const [inputs, setInputs] = useState(initial);

  // this function will allow us to update the state of the form
  const handleChange = (e) => {
    setInputs({
      // copy the existing state
      ...inputs,
      // update the state with the new data using form input name attribute
      [e.target.name]: e.target.value,
    });
  };

  // return the values of the form and the data handling functions
  return {
    inputs,
    handleChange,
  };
};

export default useForm;
