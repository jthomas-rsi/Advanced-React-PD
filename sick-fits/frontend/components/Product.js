import Link from 'next/link';
import PropTypes from 'prop-types';
import ItemStyles from './styles/ItemStyles';
import Title from './styles/Title';
import PriceTag from './styles/PriceTag';
import formatMoney from '../lib/formatMoney';
import DeleteProduct from './DeleteProduct';
import AddToCart from './AddToCart';

const Product = ({ productData }) => {
  console.log({ productData });
  return (
    <ItemStyles>
      <img
        src={productData?.photo?.image?.publicUrlTransformed}
        alt={productData.name}
      />
      <Title>
        <Link href={`/product/${productData.id}`}>{productData.name}</Link>
      </Title>
      <PriceTag>{formatMoney(productData.price)}</PriceTag>
      <p>{productData.description}</p>
      <div className="buttonList">
        <Link
          href={{
            pathname: '/update',
            query: { id: productData.id },
          }}
        >
          Edit ✏️
        </Link>
        <AddToCart id={productData.id} />
        <DeleteProduct id={productData.id}>Delete</DeleteProduct>
      </div>
    </ItemStyles>
  );
};
Product.propTypes = {
  productData: PropTypes.object.isRequired,
};

export default Product;
