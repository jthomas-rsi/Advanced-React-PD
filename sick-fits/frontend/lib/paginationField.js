import { PAGINATION_QUERY } from '../components/Pagination';

const paginationField = () => ({
  keyArgs: false, // tells Apollo we will take care of everything
  read(existing = [], { args, cache }) {
    // destructure the first and skip values from the args
    const { first, skip } = args;

    // read the number of items on the page from the cache
    const data = cache.readQuery({ query: PAGINATION_QUERY });
    const count = data?._allProductsMeta?.count;
    const page = skip / first + 1;
    const pages = Math.ceil(count / first);

    // check if we have existing items in the cache
    const items = existing.slice(skip, skip + first).filter((x) => x);

    // if there are items AND there aren't enough items to satisfy how many were requested
    // AND we are on the last page
    if (items.length && items.length !== first && page === pages) {
      return items;
    }

    // check if there are NOT items and we need to go to the network
    if (items.length !== first) {
      return false;
    }

    // if items are present return them from the cache and avoid another network call
    if (items.length) {
      // console.log(
      //   `There are ${items.length} items in the cache! Gonna send them to Apollo`
      // );
      return items;
    }

    // fallback to network request
    return false;
  },
  merge(existing, incoming, { args }) {
    const { first, skip } = args;

    // This run when the Apollo client comes back from the network with our products
    // we can then merge the two together

    const merged = existing ? existing.slice(0) : [];

    // eslint-disable-next-line no-plusplus
    for (let i = skip; i < skip + incoming.length; ++i) {
      merged[i] = incoming[i - skip];
    }

    return merged;
  },
});
export default paginationField;
