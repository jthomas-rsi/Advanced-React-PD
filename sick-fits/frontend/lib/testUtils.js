import casual from 'casual';
import { PAGINATION_QUERY } from '../components/Pagination';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  id: '6695cadeb72f324aac2a7f72',
  name: 'KITH Hoodie',
  price: 23562,
  description: 'Love this hoodie',
  photo: {
    id: '6695cadeb72f324aac2a7f71',
    image: {
      publicUrlTransformed:
        'https://res.cloudinary.com/wesbos/image/upload/v1579815935/sick-fits-keystone/5e2a13ff689b2835ae71d1a7.jpg',
      __typename: 'CloudinaryImage_File',
    },
    __typename: 'ProductImage',
  },
  __typename: 'Product',
});

const fakeUser = (overrides) => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  user: fakeUser(),
});

const fakeCartItem = (overrides) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

function makePaginationMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ];
}

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
