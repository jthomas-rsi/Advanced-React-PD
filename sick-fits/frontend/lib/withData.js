import { ApolloClient, ApolloLink, InMemoryCache } from '@apollo/client';
import { onError } from '@apollo/link-error';
import { getDataFromTree } from '@apollo/client/react/ssr';
import { createUploadLink } from 'apollo-upload-client';
import withApollo from 'next-with-apollo';
import { endpoint, prodEndpoint } from '../config';
import paginationField from './paginationField';

function createClient({ headers, initialState }) {
  return new ApolloClient({
    link: ApolloLink.from([
      // Error handling for your GraphQL requests
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors)
          graphQLErrors.forEach(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
          );
        if (networkError)
          console.log(
            `[Network error]: ${networkError}. Backend is unreachable. Is it running?`
          );
      }),
      // this uses apollo-link-http under the hood, so all the options here come from that package
      // this package allows us to do file uploads
      createUploadLink({
        uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
        fetchOptions: {
          // include credentials so we can send the cookie with the request for the logged in user
          credentials: 'include',
        },
        // pass the headers along from this request. This enables SSR with logged in state
        headers, // this is the headers from the server which allows us to render on the server side with the logged in user
      }),
    ]),
    // This data is stored in the Apollo cache. This is useful for when we want to store data in the cache and then retrieve it later
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            // TODO: We will add this together!
            allProducts: paginationField(),
          },
        },
      },
    }).restore(initialState || {}),
  });
}

export default withApollo(createClient, { getDataFromTree });
// withApollo is a higher order component that wraps our App component.
// It will make the ApolloClient available to all of our components in our app.
// It will also make sure that the data is fetched and rendered on the server before the page is served to the client. This is important for SEO and performance reasons.
