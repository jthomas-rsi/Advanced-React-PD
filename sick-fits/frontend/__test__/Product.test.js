import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import Product from '../components/Product';
import { fakeItem } from '../lib/testUtils';

const mockProduct = fakeItem();

const TestComponent = () => (
  <MockedProvider>
    <Product productData={mockProduct} />
  </MockedProvider>
);

describe('Product component ', () => {
  it('renders with correct price string', () => {
    // test using screen
    render(TestComponent());
    const h2 = screen.getByText('$235.62');
    expect(h2).toBeInTheDocument();
  });
  it('renders with link', () => {
    // test using container
    const { container } = render(TestComponent());
    const link = container.querySelector('a');
    expect(link).toHaveAttribute('href', '/product/6695cadeb72f324aac2a7f72');
    expect(link).toHaveTextContent('KITH Hoodie');
  });

  // snapshot test
  it('renders and matches the snapshot ', () => {
    const { container } = render(TestComponent());
    expect(container).toMatchSnapshot();
  });

  // alt text test - using query selector
  it('renders with correct alt text ', () => {
    const { container } = render(TestComponent());
    const img = container.querySelector('img');
    expect(img).toHaveAttribute('alt', mockProduct.name);
  });

  // image alt text test - using getByAltText
  it('renders with correct alt text ', () => {
    render(TestComponent());
    const img = screen.getByAltText(mockProduct.name);
    expect(img).toBeInTheDocument();
  });
});
