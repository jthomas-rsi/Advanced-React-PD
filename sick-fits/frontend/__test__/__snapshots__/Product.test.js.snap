// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product component  renders and matches the snapshot  1`] = `
<div>
  <div
    class="ItemStyles-sc-16pk14u-0 eAHoek"
  >
    <img
      alt="KITH Hoodie"
      src="https://res.cloudinary.com/wesbos/image/upload/v1579815935/sick-fits-keystone/5e2a13ff689b2835ae71d1a7.jpg"
    />
    <h3
      class="Title-sc-16nq74k-0 bsALs"
    >
      <a
        href="/product/6695cadeb72f324aac2a7f72"
      >
        KITH Hoodie
      </a>
    </h3>
    <span
      class="PriceTag-sc-nwbk6t-0 hxPUFn"
    >
      $235.62
    </span>
    <p>
      Love this hoodie
    </p>
    <div
      class="buttonList"
    >
      <a
        href="/update?id=6695cadeb72f324aac2a7f72"
      >
        Edit ‚úèÔ∏è
      </a>
      <button
        type="button"
      >
        Add
         to Cart üõí
      </button>
      <button
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
